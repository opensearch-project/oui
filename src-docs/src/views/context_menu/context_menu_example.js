/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import React from 'react';
import { Link } from 'react-router-dom';

import { renderToHtml } from '../../services';

import { GuideSectionTypes } from '../../components';

import {
  OuiCode,
  OuiContextMenu,
  OuiContextMenuItem,
  OuiContextMenuPanel,
} from '../../../../src/components';

import ContextMenu from './context_menu';
const contextMenuSource = require('./context_menu?raw');
const contextMenuHtml = renderToHtml(ContextMenu);
const contextMenuSnippet = `<OuiContextMenu
  initialPanelId={0}
  panels={[
    {
      id: 0,
      title: 'This is a context menu',
      items: [
        {
          name: 'Handle an onClick',
          icon: <OuiIcon type="search" size="m" />,
          onClick: () => {
            closePopover();
          },
        },
      ]
    }
  ]}
/>`;

import SinglePanel from './single_panel';
const singlePanelSource = require('./single_panel?raw');
const singlePanelHtml = renderToHtml(SinglePanel);
const singlePanelSnippet = `<OuiContextMenuPanel
  items={[
    <OuiContextMenuItem
      key=""
      onClick={}>
      This is a context menu item
    </OuiContextMenuItem>
  ]}
/>`;

import Small from './small';
const smallSizeSource = require('./small?raw');
const smallSizeHtml = renderToHtml(SinglePanel);
const smallSnippet = `<OuiContextMenuPanel
  size="s"
  items={items}
/>`;

import ContentPanel from './content_panel';
const contentPanelSource = require('./content_panel?raw');
const contentPanelHtml = renderToHtml(ContentPanel);
const contentPanelSnippet = `<OuiContextMenuPanel>
  <!-- React element as child -->
</OuiContextMenuPanel>
`;

import ContextMenuWithContent from './context_menu_with_content';
const contextMenuWithContentSource = require('./context_menu_with_content?raw');
const contextMenuWithContentHtml = renderToHtml(ContextMenuWithContent);

export const ContextMenuExample = {
  title: 'Context menu',
  sections: [
    {
      source: [
        {
          type: GuideSectionTypes.JS,
          code: contextMenuSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: contextMenuHtml,
        },
      ],
      text: (
        <p>
          <strong>OuiContextMenu</strong> is a nested menu system useful for
          navigating complicated trees. It lives within an{' '}
          <Link to="/layout/popover">
            <strong>OuiPopover</strong>
          </Link>{' '}
          which itself can be wrapped around any component (like a button in
          this example).
        </p>
      ),
      props: { OuiContextMenu, OuiContextMenuPanel, OuiContextMenuItem },
      snippet: contextMenuSnippet,
      demo: <ContextMenu />,
    },
    {
      title: 'Sizes',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: smallSizeSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: smallSizeHtml,
        },
      ],
      text: (
        <p>
          <strong>OuiContextMenu</strong> supports a small and medium{' '}
          <OuiCode>size</OuiCode>. The default size is medium,{' '}
          <OuiCode>m</OuiCode>, and should be used for most menus and major
          actions such as top application menus. Use the smaller size,{' '}
          <OuiCode>s</OuiCode>, for a more compressed version containing minor
          actions or repeated menus like in <strong>OuiTable</strong>{' '}
          pagination.
        </p>
      ),
      snippet: smallSnippet,
      demo: <Small />,
    },
    {
      title: 'With single panel',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: singlePanelSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: singlePanelHtml,
        },
      ],
      text: (
        <p>
          Use <strong>OuiContextMenuPanel</strong> for simple, non-nested
          context menus. The below pagination example has no nesting and no
          title.
        </p>
      ),
      snippet: singlePanelSnippet,
      demo: <SinglePanel />,
    },
    {
      title: 'Displaying custom elements',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: contentPanelSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: contentPanelHtml,
        },
      ],
      text: (
        <p>
          If you have custom content to show instead of a list of options, you
          can pass a React element as a child to{' '}
          <strong>OuiContextMenuPanel</strong>.
        </p>
      ),
      snippet: contentPanelSnippet,
      demo: <ContentPanel />,
    },
    {
      title: 'Using panels with mixed items & content',
      source: [
        {
          type: GuideSectionTypes.JS,
          code: contextMenuWithContentSource,
        },
        {
          type: GuideSectionTypes.HTML,
          code: contextMenuWithContentHtml,
        },
      ],
      text: (
        <div>
          <p>
            Context menu panels can be passed React elements through the{' '}
            <OuiCode>content</OuiCode> prop instead of <OuiCode>items</OuiCode>.
            The panel will display your custom content without modification.
          </p>
          <p>
            If your panel contents have different widths or you need to ensure
            that a specific context menu panel has a certain width, add{' '}
            <OuiCode language="ts">width: [number of pixels]</OuiCode> to the
            panel tree.
          </p>
          <p>
            You can add separator lines in the <OuiCode>items</OuiCode> prop if
            you define an item as{' '}
            <OuiCode language="ts">{'{isSeparator: true}'}</OuiCode>. This will
            pass the rest of its fields as props to a{' '}
            <Link to="/layout/horizontal-rule">
              <strong>OuiHorizontalRule</strong>
            </Link>{' '}
            component.
          </p>
        </div>
      ),
      demo: <ContextMenuWithContent />,
    },
  ],
};
