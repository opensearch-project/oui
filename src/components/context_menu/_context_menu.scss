/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

$ouiContextMenuWidth: $ouiSize * 16;

.ouiContextMenu {
  width: $ouiContextMenuWidth;
  max-width: 100%;
  position: relative;
  overflow: hidden;
  transition: height $ouiAnimSpeedFast $ouiAnimSlightResistance;
  border-radius: $ouiBorderRadius;

  .ouiContextMenu__content {
    padding: $ouiSizeS;
  }
}

/**
  * 1. When there are multiple ContextMenuPanels, the ContextMenu will absolutely
  *    position them. ContextMenuPanel will break the layout of a Popover if it's
  *    absolutely positioned by default.
  */
.ouiContextMenu__panel {
  position: absolute; /* 1 */
}

.ouiContextMenu__icon {
  margin-right: $ouiSizeS;
}


/* OUI -> EUI Aliases */
$euiContextMenuWidth: $ouiContextMenuWidth;

/* End of Aliases */
