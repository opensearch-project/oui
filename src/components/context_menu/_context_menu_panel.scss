/*!
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

.ouiContextMenuPanel {
  width: 100%;
  visibility: visible;
  outline-offset: -$ouiFocusRingSize;

  &:focus {
    outline: none; // Hide focus ring because of `tabindex=-1` on Safari
  }

  &.ouiContextMenuPanel-txInLeft {
    pointer-events: none;
    animation: ouiContextMenuPanelTxInLeft $ouiAnimSpeedNormal $ouiAnimSlightResistance;
  }

  &.ouiContextMenuPanel-txOutLeft {
    pointer-events: none;
    animation: ouiContextMenuPanelTxOutLeft $ouiAnimSpeedNormal $ouiAnimSlightResistance;
  }

  &.ouiContextMenuPanel-txInRight {
    pointer-events: none;
    animation: ouiContextMenuPanelTxInRight $ouiAnimSpeedNormal $ouiAnimSlightResistance;
  }

  &.ouiContextMenuPanel-txOutRight {
    pointer-events: none;
    animation: ouiContextMenuPanelTxOutRight $ouiAnimSpeedNormal $ouiAnimSlightResistance;
  }
}

.ouiContextMenuPanel--next {
  transform: translateX($ouiContextMenuWidth);
  visibility: hidden;
}

.ouiContextMenuPanel--previous {
  transform: translateX(-$ouiContextMenuWidth);
  visibility: hidden;
}

.ouiContextMenuPanelTitle {
  @include ouiPopoverTitle;
  padding: $ouiSizeM;
  width: 100%;
  text-align: left;
  outline-offset: -$ouiFocusRingSize;

  &:enabled:hover,
  &:enabled:focus {
    text-decoration: underline;
  }

  &--small {
    @include ouiPopoverTitle;
    padding: ($ouiSizeS * .75) $ouiSizeS;
  }
}

@keyframes ouiContextMenuPanelTxInLeft {
  0% {
    transform: translateX($ouiContextMenuWidth);
  }

  100% {
    transform: translateX(0);
  }
}

@keyframes ouiContextMenuPanelTxOutLeft {
  0% {
    transform: translateX(0);
  }

  100% {
    transform: translateX(-$ouiContextMenuWidth);
  }
}

@keyframes ouiContextMenuPanelTxInRight {
  0% {
    transform: translateX(-$ouiContextMenuWidth);
  }

  100% {
    transform: translateX(0);
  }
}

@keyframes ouiContextMenuPanelTxOutRight {
  0% {
    transform: translateX(0);
  }

  100% {
    transform: translateX($ouiContextMenuWidth);
  }
}
