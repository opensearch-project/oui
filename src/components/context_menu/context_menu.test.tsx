/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React from 'react';
import { render, screen, act } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { requiredProps } from '../../test';

import { OuiContextMenu, SIZES } from './context_menu';
import { setTimeout } from 'timers';

const panel3 = {
  id: 3,
  title: '3',
  items: [
    {
      name: <span style={{ color: 'tomato' }}>foo</span>,
      key: 'foo',
    },
  ],
};

const panel2 = {
  id: 2,
  title: '2',
  content: <div>2</div>,
};

const panel1 = {
  id: 1,
  title: '1',
  width: 400,
  items: [
    {
      name: '2a',
      panel: 2,
    },
    {
      name: '2b',
      panel: 2,
    },
    {
      name: '2c',
      panel: 2,
    },
  ],
};

const panel0 = {
  id: 0,
  title: '0',
  items: [
    {
      name: '1',
      panel: 1,
    },
  ],
};

const panels = [panel0, panel1, panel2, panel3];

export const tick = (ms = 0) =>
  new Promise((resolve) => {
    setTimeout(resolve, ms);
  });

describe('OuiContextMenu', () => {
  test('is rendered', () => {
    const { container } = render(<OuiContextMenu {...requiredProps} />);

    expect(container).toMatchSnapshot();
  });

  it('panel item can contain JSX', () => {
    const { container } = render(
      <OuiContextMenu panels={panels} initialPanelId={3} />
    );

    expect(container).toMatchSnapshot();
  });

  it('panel item can be a separator line', () => {
    const { container } = render(
      <OuiContextMenu
        panels={[
          {
            id: 3,
            title: 'Testing separator',
            items: [
              { name: 'Foo', key: 'foo' },
              { isSeparator: true },
              { name: 'Bar', key: 'bar' },
            ],
          },
        ]}
        initialPanelId={3}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it('can pass-through horizontal rule props', () => {
    const { container } = render(
      <OuiContextMenu
        panels={[
          {
            id: 3,
            title: 'Testing separator',
            items: [
              {
                isSeparator: true,
                key: 'separator',
                margin: 's',
                size: 'half',
              },
            ],
          },
        ]}
        initialPanelId={3}
      />
    );

    expect(container).toMatchSnapshot();
  });

  describe('props', () => {
    describe('panels and initialPanelId', () => {
      it('renders the referenced panel', () => {
        const { container } = render(
          <OuiContextMenu panels={panels} initialPanelId={2} />
        );

        expect(container).toMatchSnapshot();
      });

      it('allows you to click the title button to go back to the previous panel', async () => {
        const { container } = render(
          <OuiContextMenu panels={panels} initialPanelId={2} />
        );

        await tick(20);

        expect(container).toMatchSnapshot();

        // Navigate to a different panel.
        const user = userEvent.setup();
        const titleButton = screen.getByTestId('contextMenuPanelTitleButton');

        await act(async () => {
          await user.click(titleButton);
          await tick(20);
        });

        expect(container).toMatchSnapshot();
      });
    });

    describe('size', () => {
      SIZES.forEach((size) => {
        it(`${size} is rendered`, () => {
          const { container } = render(
            <OuiContextMenu panels={panels} initialPanelId={2} size={size} />
          );

          expect(container).toMatchSnapshot();
        });
      });
    });
  });
});
