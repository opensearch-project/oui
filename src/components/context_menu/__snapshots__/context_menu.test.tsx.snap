// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OuiContextMenu can pass-through horizontal rule props 1`] = `
<div>
  <div
    class="ouiContextMenu"
    style="height: 0px;"
  >
    <div
      class="ouiContextMenuPanel ouiContextMenu__panel"
      tabindex="-1"
    >
      <div
        class="ouiContextMenuPanelTitle"
      >
        <span
          class="ouiContextMenu__itemLayout"
        >
          Testing separator
        </span>
      </div>
      <div>
        <div>
          <hr
            class="ouiHorizontalRule ouiHorizontalRule--half ouiHorizontalRule--marginSmall"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`OuiContextMenu is rendered 1`] = `
<div>
  <div
    aria-label="aria-label"
    class="ouiContextMenu testClass1 testClass2"
    data-test-subj="test subject string"
  />
</div>
`;

exports[`OuiContextMenu panel item can be a separator line 1`] = `
<div>
  <div
    class="ouiContextMenu"
    style="height: 0px;"
  >
    <div
      class="ouiContextMenuPanel ouiContextMenu__panel"
      tabindex="-1"
    >
      <div
        class="ouiContextMenuPanelTitle"
      >
        <span
          class="ouiContextMenu__itemLayout"
        >
          Testing separator
        </span>
      </div>
      <div>
        <div>
          <button
            class="ouiContextMenuItem"
            type="button"
          >
            <span
              class="ouiContextMenu__itemLayout"
            >
              <span
                class="ouiContextMenuItem__text"
              >
                Foo
              </span>
            </span>
          </button>
          <hr
            class="ouiHorizontalRule ouiHorizontalRule--full"
          />
          <button
            class="ouiContextMenuItem"
            type="button"
          >
            <span
              class="ouiContextMenu__itemLayout"
            >
              <span
                class="ouiContextMenuItem__text"
              >
                Bar
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`OuiContextMenu panel item can contain JSX 1`] = `
<div>
  <div
    class="ouiContextMenu"
    style="height: 0px;"
  >
    <div
      class="ouiContextMenuPanel ouiContextMenu__panel"
      tabindex="-1"
    >
      <div
        class="ouiContextMenuPanelTitle"
      >
        <span
          class="ouiContextMenu__itemLayout"
        >
          3
        </span>
      </div>
      <div>
        <div>
          <button
            class="ouiContextMenuItem"
            type="button"
          >
            <span
              class="ouiContextMenu__itemLayout"
            >
              <span
                class="ouiContextMenuItem__text"
              >
                <span
                  style="color: tomato;"
                >
                  foo
                </span>
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`OuiContextMenu props panels and initialPanelId allows you to click the title button to go back to the previous panel 1`] = `
<div>
  <div
    class="ouiContextMenu"
    style="height: 0px;"
  >
    <div
      class="ouiContextMenuPanel ouiContextMenu__panel"
      tabindex="-1"
    >
      <button
        class="ouiContextMenuPanelTitle"
        data-test-subj="contextMenuPanelTitleButton"
        type="button"
      >
        <span
          class="ouiContextMenu__itemLayout"
        >
          <span
            class="ouiContextMenu__icon"
            data-ouiicon-type="arrowLeft"
          />
          <span
            class="ouiContextMenu__text"
          >
            2
          </span>
        </span>
      </button>
      <div>
        <div>
          <div>
            2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`OuiContextMenu props panels and initialPanelId allows you to click the title button to go back to the previous panel 2`] = `
<div>
  <div
    class="ouiContextMenu"
    style="height: 0px; width: 400px;"
  >
    <div
      class="ouiContextMenuPanel ouiContextMenu__panel ouiContextMenuPanel-txOutRight"
      tabindex="-1"
    >
      <button
        class="ouiContextMenuPanelTitle"
        data-test-subj="contextMenuPanelTitleButton"
        type="button"
      >
        <span
          class="ouiContextMenu__itemLayout"
        >
          <span
            class="ouiContextMenu__icon"
            data-ouiicon-type="arrowLeft"
          />
          <span
            class="ouiContextMenu__text"
          >
            2
          </span>
        </span>
      </button>
      <div>
        <div>
          <div>
            2
          </div>
        </div>
      </div>
    </div>
    <div
      class="ouiContextMenuPanel ouiContextMenu__panel ouiContextMenuPanel-txInRight"
      tabindex="-1"
    >
      <button
        class="ouiContextMenuPanelTitle"
        data-test-subj="contextMenuPanelTitleButton"
        type="button"
      >
        <span
          class="ouiContextMenu__itemLayout"
        >
          <span
            class="ouiContextMenu__icon"
            data-ouiicon-type="arrowLeft"
          />
          <span
            class="ouiContextMenu__text"
          >
            1
          </span>
        </span>
      </button>
      <div>
        <div>
          <button
            class="ouiContextMenuItem"
            type="button"
          >
            <span
              class="ouiContextMenu__itemLayout"
            >
              <span
                class="ouiContextMenuItem__text"
              >
                2a
              </span>
              <span
                class="ouiContextMenu__arrow"
                data-ouiicon-type="arrowRight"
              />
            </span>
          </button>
          <button
            class="ouiContextMenuItem"
            type="button"
          >
            <span
              class="ouiContextMenu__itemLayout"
            >
              <span
                class="ouiContextMenuItem__text"
              >
                2b
              </span>
              <span
                class="ouiContextMenu__arrow"
                data-ouiicon-type="arrowRight"
              />
            </span>
          </button>
          <button
            class="ouiContextMenuItem"
            type="button"
          >
            <span
              class="ouiContextMenu__itemLayout"
            >
              <span
                class="ouiContextMenuItem__text"
              >
                2c
              </span>
              <span
                class="ouiContextMenu__arrow"
                data-ouiicon-type="arrowRight"
              />
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`OuiContextMenu props panels and initialPanelId renders the referenced panel 1`] = `
<div>
  <div
    class="ouiContextMenu"
    style="height: 0px;"
  >
    <div
      class="ouiContextMenuPanel ouiContextMenu__panel"
      tabindex="-1"
    >
      <button
        class="ouiContextMenuPanelTitle"
        data-test-subj="contextMenuPanelTitleButton"
        type="button"
      >
        <span
          class="ouiContextMenu__itemLayout"
        >
          <span
            class="ouiContextMenu__icon"
            data-ouiicon-type="arrowLeft"
          />
          <span
            class="ouiContextMenu__text"
          >
            2
          </span>
        </span>
      </button>
      <div>
        <div>
          <div>
            2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`OuiContextMenu props size m is rendered 1`] = `
<div>
  <div
    class="ouiContextMenu"
    style="height: 0px;"
  >
    <div
      class="ouiContextMenuPanel ouiContextMenu__panel"
      tabindex="-1"
    >
      <button
        class="ouiContextMenuPanelTitle"
        data-test-subj="contextMenuPanelTitleButton"
        type="button"
      >
        <span
          class="ouiContextMenu__itemLayout"
        >
          <span
            class="ouiContextMenu__icon"
            data-ouiicon-type="arrowLeft"
          />
          <span
            class="ouiContextMenu__text"
          >
            2
          </span>
        </span>
      </button>
      <div>
        <div>
          <div>
            2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`OuiContextMenu props size s is rendered 1`] = `
<div>
  <div
    class="ouiContextMenu ouiContextMenu--small"
    style="height: 0px;"
  >
    <div
      class="ouiContextMenuPanel ouiContextMenu__panel"
      tabindex="-1"
    >
      <button
        class="ouiContextMenuPanelTitle ouiContextMenuPanelTitle--small"
        data-test-subj="contextMenuPanelTitleButton"
        type="button"
      >
        <span
          class="ouiContextMenu__itemLayout"
        >
          <span
            class="ouiContextMenu__icon"
            data-ouiicon-type="arrowLeft"
          />
          <span
            class="ouiContextMenu__text"
          >
            2
          </span>
        </span>
      </button>
      <div>
        <div>
          <div>
            2
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
